lib/ae_mdw/contract.ex

```diff
@@ -451,18 +452,32 @@ defmodule AeMdw.Contract do
       trees_in = apply(consensus, :state_pre_transform_micro_node, [node, trees_in])
       env = :aetx_env.tx_env_from_key_header(prev_key_header, prev_key_hash, time, prev_hash)
 
-      {:ok, _, _, events} = :aec_block_micro_candidate.apply_block_txs_strict(txs, trees_in, env)
+      {:ok, _, _, events} =
+        :aec_block_micro_candidate.apply_block_txs_strict(txs_taken, trees_in, env)
+
       events
     else
       []
     end
   end
 
-  defp has_contract_tx?(mb_txs) do
-    Enum.any?(mb_txs, fn signed_tx ->
-      {mod, _tx} = :aetx.specialize_callback(:aetx_sign.tx(signed_tx))
-      mod.type() in [:contract_create_tx, :contract_call_tx]
-    end)
+  defp txs_until_last_contract_tx(mb_txs) do
+    {tx_pos, last_ct_tx_pos} =
+      Enum.reduce(mb_txs, {1, -1}, fn signed_tx, {tx_pos, last_index} ->
```

Comment:
traverse the whole list of txs but compensates and it was already done most of the cases with `Enum.any?`